src/dataflow-types/src/client/controller/compute.rs

```diff
@@ -199,21 +228,81 @@ impl<'a, C: Client<T>, T: Timestamp + Lattice> ComputeController<'a, C, T> {
             ))
             .await;
     }
-    pub async fn allow_index_compaction(
-        &mut self,
-        frontiers: Vec<(GlobalId, Antichain<T>)>,
-    ) -> Result<(), ComputeError> {
-        for (id, frontier) in frontiers.iter() {
-            self.compute.since_uppers.advance_since_for(*id, frontier);
-        }
+
+    pub async fn allow_index_compaction(&mut self, frontiers: Vec<(GlobalId, Antichain<T>)>) {
+        // The coordinator currently sends compaction commands for identifiers that do not exist.
+        // Until that changes, we need to be oblivious to errors, or risk not compacting anything.
+
+        // // Validate that the ids exist.
+        // self.validate_ids(frontiers.iter().map(|(id, _)| *id))?;
+        //
+        // // Downgrade the implicit capability for each referenced id.
+        // for (id, frontier) in frontiers.iter() {
+        //     // Apply the updates but ignore the results for now.
+        //     // TODO(mcsherry): observe the results and allow compaction.
+        //     let _ = self
+        //         .collection_mut(*id)?
+        //         .capability_downgrade(frontier.clone());
+        // }
+        // // TODO(mcsherry): Delay compation subject to read capability constraints.
 
         self.client
             .send(Command::Compute(
                 ComputeCommand::AllowIndexCompaction(frontiers),
                 self.instance,
             ))
             .await;
+    }
+}
 
+// Internal interface
+impl<'a, C: Client<T>, T: Timestamp + Lattice> ComputeController<'a, C, T> {
+    pub fn collection_mut(
+        &mut self,
+        id: GlobalId,
+    ) -> Result<&mut CollectionState<T>, ComputeError> {
+        self.compute
+            .collections
+            .get_mut(&id)
+            .ok_or(ComputeError::IdentifierMissing(id))
+    }
+    pub fn validate_ids(&self, ids: impl Iterator<Item = GlobalId>) -> Result<(), ComputeError> {
+        for id in ids {
+            self.collection(id)?;
+        }
         Ok(())
     }
 }
+
+/// State maintained about individual collections.
+pub struct CollectionState<T> {
+    /// Accumulation of read capabilities for the collection.
+    pub(super) since: MutableAntichain<T>,
+    /// Reported progress in the write capabilities.
+    pub(super) upper_frontier: MutableAntichain<T>,
```

Comment:
for consistency can we change these to (`since_frontier`/`upper_frontier`) or (`since`/`upper`)?

Reply:
They are different! `since` should perhaps be `since_capability`, though. The first is a held read capability, and the second (`upper_frontier`) is what has been reported about the held write capabilities (though is not a held write capability).

Reply:
I'll add a bunch of better comments here in a bit. The `since` thing is about to change with more flexible capabilities arriving.